File: design-pattern/streategy-pattern/streategy-pattern.ts

//interface for the walkable streategy
interface walkAble {
    walk(): void;
};

//concreate bheviour
class noWalkRobot implements walkAble {
    walk(): void {
        console.log("no walk robot");
    }
};


class walkRobot implements walkAble {
    walk(): void {
        console.log('this robot can walk')
    }
};

interface talkAble {
    talk(): void;
};


class noTalkRobot implements talkAble {
    talk(): void {
        console.log("no talk robot");
    }
};


class talkRobot implements talkAble {
    talk(): void {
        console.log('robot is talking')
    }
};


class Robot {
    private walkAble: walkAble;
    private talkAble: talkAble;

    constructor(walk: walkAble, talk: talkAble) {
        this.walkAble = walk;
        this.talkAble = talk;
    };

    walk(): void {
        this.walkAble.walk();
    }

    talk(): void {
        this.talkAble.talk();
    }
};

const robotR1 = new Robot(new noWalkRobot(),
    new noTalkRobot())
robotR1.walk();
robotR1.talk();

File: design-pattern/streategy-pattern/streategy-pattern.cpp
#include <iostream>

using namespace std;

// abstract streategy
class walkableRobot
{
public:
    virtual void walk() = 0;
    virtual ~walkableRobot() {};
};

class normalwalk : public walkableRobot
{
public:
    void walk() override
    {
        cout << "normal robot" << endl;
    }
};

class nowalk : public walkableRobot
{
public:
    void walk() override
    {
        cout << "robot is not walking " << endl;
    }
};

class talkableRobot
{
public:
    virtual void talk() = 0;
    virtual ~talkableRobot() {};
};

class normalTalk : public talkableRobot
{
public:
    void talk()
    {
        cout << "normal robot is talking" << endl;
    }
};

class noTalk : public talkableRobot
{
public:
    void talk()
    {
        cout << "robot is not talking" << endl;
    }
};

class Robot
{
protected:
    walkableRobot *walkRobot;
    talkableRobot *talkRobot;

public:
    Robot(walkableRobot *walk, talkableRobot *talk)
    {
        this->talkRobot = talk;
        this->walkRobot = walk;
    };

    void walk()
    {
        walkRobot->walk();
    }

    void talk()
    {
        talkRobot->talk();
    }
};

int main(void)
{
    // dynamicaly changing the behaviour
    Robot *r1 = new Robot(new nowalk(), new normalTalk());

    r1->talk();
    r1->walk();

    return 0;
};